system/doc/efficiency_guide/maps.xml

```diff
@@ -0,0 +1,647 @@
+
+      <p>When the value for an existing key is updated, the key tuple is not
+      be updated, allowing the key tuple to be shared with instances of the
+      map that have the same keys. In fact, the key tuple can be shared between all
+      instances of a map with some care. To arrange that, define a function that
+      returns a map. For example:</p>
+
+      <code type="erl"><![CDATA[
+new() ->
+    #{a => default, b => default, c => default}.]]></code>
+
+      <p>Defined like this, the key tuple <c>{a,b,c}</c> will be a
+      global literal. To ensure that the key tuple is shared
+      when creating an instance of the map, always call
+      <c>new()</c> and modify the returned map:</p>
+
+      <code type="erl"><![CDATA[
+    (SOME_MODULE:new())#{a := 42}.]]></code>
+
+      <p>Using the map syntax with small maps is particularly efficient.
+      As long as the keys are known at compile-time, the map is updated
+      in one go, making the time to update a map essentially constant regardless
+      of the number of keys updated. The same goes for matching.</p>
+
+      <p>The memory size for a small map is the size of all keys and values
+      plus 5 words. See <seeguide marker="advanced#memory">Advanced</seeguide>
+      for more information about memory sizes.</p>
+    </section>
+
+    <section>
+      <title>How Large Maps are Implemented</title>
+
+      <p>A map with more than 32 elements is implemented as a <url
+      href="https://en.wikipedia.org/wiki/Hash_array_mapped_trie">Hash
+      array mapped trie (HAMT)</url>. A large map can be efficiently
+      searched and updated regardless of the number of elements in the
+      map.</p>
+
+      <p>There is less performance to be gained by matching or
+      updating multiple elements using the map syntax on a large map
+      compared to a small map. The execution time is roughly proportional
+      to the number of elements matched or updated.</p>
+
+      <p>The storage overhead for a large map is higher than for a
+      small map. For a large map, the extra number of words besides
+      the keys and values is roughly proportional to the number of
+      elements. For a map with 33 elements the overhead is at
+      least 53 heap words according to the formula in <seeguide
+      marker="advanced#memory">Advanced</seeguide> (compared to 5
+      extra words for a small map regardless of the number of
+      elements).</p>
+
+      <p>When a large map is updated, the updated map and the original
+      map will share common parts of the HAMT, but sharing will never
+      be as effective as the sharing of the key tuple for a small
+      maps.</p>
+
+      <p>Therefore, if maps are used instead of records and it is
+      expected that many instances of the map will be created, it is
+      more efficient from a memory standpoint to avoid using large
+      maps (for example, by grouping related map elements into sub
+      maps to reduce the number of elements).</p>
```

Comment:
One thing that might be good to mention is that the "template" map optimization is not the most memory efficient  representation if most "fields" has their default value. For example a map with 30 fields where only one field does not have the default value would use more memory with the "template" map optimization that without this optimization.

Reply:
The template trick is all only meant to share the key tuple, not the array of values. The key tuple will be shared.

Reply:
Yes. You are right. I think my comment is misplaced. We recommend the template trick partly because it should reduce memory consumption. The point I was trying to make is that the template trick does not always reduce memory consumption so perhaps it can be useful to mention that.